File: bin/bdfcheck.js
/*
  Copyright (C) 2017-2020 Dimitar Toshkov Zhekov <dimitar.zhekov@gmail.com>

  This program is free software; you can redistribute it and/or modify it
  under the terms of the GNU General Public License as published by the Free
  Software Foundation; either version 2 of the License, or (at your option)
  any later version.

  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
  or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
  for more details.

  You should have received a copy of the GNU General Public License along
  with this program; if not, write to the Free Software Foundation, Inc.,
  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

'use strict';

const fnutil = require('./fnutil.js');
const fncli = require('./fncli.js');
const fnio = require('./fnio.js');
const bdf = require('./bdf.js');

// -- Params --
class Params extends fncli.Params {
	constructor() {
		super();
		this.asciiChars = true;
		this.bbxExceeds = true;
		this.duplCodes = -1;
		this.extraBits = true;
		this.attributes = true;
		this.duplNames = -1;
		this.duplProps = true;
		this.commonSlant = true;
		this.commonWeight = true;
		this.xlfdFontNm = true;
		this.yWidthZero = true;
	}
}

// -- Options --
const HELP = ('' +
	'usage: bdfcheck [options] [INPUT...]\n' +
	'Check BDF font(s) for various problems\n' +
	'\n' +
	'  -A          disable non-ascii characters check\n' +
	'  -B          disable BBX exceeding FONTBOUNDINGBOX checks\n' +
	'  -c/-C       enable/disable duplicate character codes check\n' +
	'              (default = enabled for registry ISO10646)\n' +
	'  -E          disable extra bits check\n' +
	'  -I          disable ATTRIBUTES check\n' +
	'  -n/-N       enable duplicate character names check\n' +
	'              (default = enabled for registry ISO10646)\n' +
	'  -P          disable duplicate properties check\n' +
	'  -S          disable common slant check\n' +
	'  -W          disable common weight check\n' +
	'  -X          disable XLFD font name check\n' +
	'  -Y          disable zero WIDTH Y check\n' +
	'  --help      display this help and exit\n' +
	'  --version   display the program version and license, and exit\n' +
	'  --excstk    display the exception stack on error\n' +
	'\n' +
	'File directives: COMMENT bdfcheck --enable|disable-<check-name>\n' +
	'  (also available as long command line options)\n' +
	'\n' +
	'Check names: ascii-chars, bbx-exceeds, duplicate-codes, extra-bits,\n' +
	'  attributes, duplicate-names, duplicate-properties, common-slant,\n' +
	'  common-weight, xlfd-font, ywidth-zero\n' +
	'\n' +
	'The input BDF(s) must be v2.1 with unicode encoding.\n');

const VERSION = 'bdfcheck 1.61, Copyright (C) 2017-2020 Dimitar Toshkov Zhekov\n\n' + fnutil.GPL2PLUS_LICENSE;

class Options extends fncli.Options {
	constructor() {
		super([], HELP, VERSION);
	}

	parse(name, directive, params) {
		const value = name.startsWith('--enable') || name[1].match('[a-z]');

		switch (name) {
		case '-A':
		case '--enable-ascii-chars':
		case '--disable-ascii-chars':
			params.asciiChars = value;
			break;
		case '-B':
		case '--enable-bbx-exceeds':
		case '--disable-bbx-exceeds':
			params.bbxExceeds = value;
			break;
		case '-c':
		case '-C':
		case '--enable-duplicate-codes':
		case '--disable-duplicate-codes':
			params.duplCodes = value;
			break;
		case '-E':
		case '--enable-extra-bits':
		case '--disable-extra-bits':
			params.extraBits = value;
			break;
		case '-I':
		case '--enable-attributes':
		case '--disable-attributes':
			params.attributes = value;
			break;
		case '-n':
		case '-N':
		case '--enable-duplicate-names':
		case '--disable-duplicate-names':
			params.duplNames = value;
			break;
		case '-P':
		case '--enable-duplicate-properties':
		case '--disable-duplicate-properties':
			params.duplProps = value;
			break;
		case '-S':
		case '--enable-common-slant':
		case '--disable-common-slant':
			params.commonSlant = value;
			break;
		case '-W':
		case '--enable-common-weight':
		case '--disable-common-weight':
			params.commonWeight = value;
			break;
		case '-X':
		case '--enable-xlfd-font':
		case '--disable-xlfd-font':
			params.xlfdFontNm = value;
			break;
		case '-Y':
		case '--enable-ywidth-zero':
		case '--disable-ywidth-zero':
			params.yWidthZero = value;
			break;
		default:
			return directive !== true && this.fallback(name, params);
		}

		return directive !== true || name.startsWith('--');
	}
}

// -- DupMap --
class DupMap extends Map {
	constructor(prefix, descript, severity) {
		super();
		this.prefix = prefix;
		this.descript = descript;
		this.severity = severity;
	}

	check() {
		this.forEach((lines, value) => {
			if (lines.length > 1) {
				let text = `duplicate ${this.descript} ${value} at lines`;

				for (let index = 0; index < lines.length; index++) {
					text += (index === 0 ? ' ' : index === lines.length - 1 ? ' and ' : ', ');
					text += lines[index];
				}
				fnutil.message(this.prefix, this.severity, text);
			}
		});
	}

	push(value, lineNo) {
		let lines = this.get(value);

		if (lines != null) {
			lines.push(lineNo);
		} else {
			this.set(value, [lineNo]);
		}
	}
}

// -- InputFileStream --
const MODE = Object.freeze({
	META: 0,
	PROPS: 1,
	BITMAP: 2
});

class InputFileStream extends fnio.InputFileStream {
	constructor(fileName, parsed) {
		super(fileName);
		this.parsed = parsed;
		this.mode = MODE.META;
		this.proplocs = new DupMap(this.location(), 'property');
		this.namelocs = new DupMap(this.location(), 'character name', 'warning');
		this.codelocs = new DupMap(this.location(), 'encoding', 'warning');
		this.HANDLERS = [
			[ 'STARTCHAR',       value => this.appendName(value) ],
			[ 'ENCODING',        value => this.appendCode(value) ],
			[ 'SWIDTH',          value => this.checkWidth('SWIDTH', value, bdf.Width.parseS) ],
			[ 'DWIDTH',          value => this.checkWidth('DWIDTH', value, bdf.Width.parseD) ],
			[ 'BBX',             value => this.setLastBox(value) ],
			[ 'BITMAP',          () => this.setMode(MODE.BITMAP) ],
			[ 'SIZE',            InputFileStream.checkSize ],
			[ 'ATTRIBUTES',      value => this.checkAttr(value) ],
			[ 'STARTPROPERTIES', () => this.setMode(MODE.PROPS) ],
			[ 'FONTBOUNDINGBOX', value => this.setFontBox(value) ]
		];
		this.xlfdName = false;
		this.lastBox = null;
		this.fontBox = null;
		this.options = new Options();
	}

	append(option, valocs, value) {
		if (option) {
			valocs.push(value, this.lineNo);
		}
	}

	appendCode(value) {
		fnutil.parseDec('encoding', value);
		this.append(this.parsed.duplCodes, this.codelocs, value);
	}

	appendName(value) {
		this.append(this.parsed.duplNames, this.namelocs, `"${value}"`);
	}

	checkWidth(name, value, parse) {
		if (this.parsed.yWidthZero && parse(name, value).y !== 0) {
			fnutil.warning(this.location(), `non-zero ${name} Y`);
		}
	}

	setFontBox(value) {
		this.fontBox = bdf.BBX.parse('FONTBOUNDINGBOX', value);
	}

	setLastBox(value) {
		const bbx = bdf.BBX.parse('BBX', value);

		if (this.parsed.bbxExceeds) {
			let exceeds = [];

			if (bbx.xoff < this.fontBox.xoff) {
				exceeds.push('xoff < FONTBOUNDINGBOX xoff');
			}
			if (bbx.yoff < this.fontBox.yoff) {
				exceeds.push('yoff < FONTBOUNDINGBOX yoff');
			}
			if (bbx.width > this.fontBox.width) {
				exceeds.push('width > FONTBOUNDINGBOX width');
			}
			if (bbx.height > this.fontBox.height) {
				exceeds.push('height > FONTBOUNDINGBOX height');
			}
			exceeds.forEach(exceed => {
				fnutil.message(this.location(), '', exceed);
			});
		}
		this.lastBox = bbx;
	}

	setMode(newMode) {
		this.mode = newMode;
	}

	static checkSize(value) {
		const words = fnutil.splitWords('SIZE', value, 3);

		fnutil.parseDec('point size', words[0], 1, null);
		fnutil.parseDec('x resolution', words[1], 1, null);
		fnutil.parseDec('y resolution', words[2], 1, null);
	}

	checkAttr(value) {
		if (!value.match(/^[\dA-Fa-f]{4}$/)) {
			throw new Error('ATTRIBUTES must be 4 hex-encoded characters');
		}
		if (this.parsed.attributes) {
			fnutil.warning(this.location(), 'ATTRIBUTES may cause problems with freetype');
		}
	}

	checkFont(value) {
		const xlfd = value.substring(4).trimLeft().split('-', 16);

		if (xlfd.length === 15 && xlfd[0] === '') {
			let unicode = (xlfd[bdf.XLFD.CHARSET_REGISTRY].toUpperCase() === 'ISO10646');

			if (this.parsed.duplCodes === -1) {
				this.parsed.duplCodes = unicode;
			}
			if (this.parsed.duplNames === -1) {
				this.parsed.duplNames = unicode;
			}

			if (this.parsed.commonWeight) {
				let weight = xlfd[bdf.XLFD.WEIGHT_NAME];
				let compare = weight.toLowerCase();
				let consider = compare.includes('bold') ? 'Bold' : 'Normal';

				if (compare === 'medium' || compare === 'regular') {
					compare = 'normal';
				}
				if (compare !== consider.toLowerCase()) {
					fnutil.warning(this.location(), `weight "${weight}" may be considered ${consider}`);
				}
			}

			if (this.parsed.commonSlant) {
				let slant = xlfd[bdf.XLFD.SLANT];
				let consider = slant.match(/^[IO]/) ? 'Italic' : 'Regular';

				if (slant.match(/^[IOR]$/) == null) {
					fnutil.warning(this.location(), `slant "${slant}" may be considered ${consider}`);
				}
			}
		} else {
			if (this.parsed.xlfdFontNm) {
				fnutil.warning(this.location(), 'non-XLFD font name');
			}
			value = 'FONT --------------';
		}

		return value;
	}

	checkProp(line) {
		const match = line.match(/^(\w+)\s+([-\d"].*)$/);

		if (match == null) {
			throw new Error('invalid property format');
		}

		const name = match[1];
		const value = match[2];

		if (value.startsWith('"')) {
			if (value.length < 2 || !value.endsWith('"')) {
				throw new Error('no closing double quote');
			}
			if (value.substring(1, value.length - 1).match(/[^"]"[^"]/)) {
				throw new Error('unescaped double quote');
			}
		} else {
			fnutil.parseDec('value', value, null, null);
		}

		this.append(this.parsed.duplProps, this.proplocs, name);
		return `P${this.lineNo} 1`;
	}

	checkBitmap(line) {
		if (line.length !== this.lastBox.rowSize() * 2) {
			throw new Error('invalid bitmap length');
		} else if (line.match(/^[\dA-Fa-f]+$/) == null) {
			throw new Error('invalid bitmap data');
		} else if (this.parsed.extraBits) {
			const data = Buffer.from(line, 'hex');
			const checkX = (this.lastBox.width - 1) | 7;
			const lastByte = data[data.length - 1];
			let bitNo = 7 - (this.lastBox.Width & 7);

			for (let x = this.lastBox.Width; x <= checkX; x++) {
				if (lastByte & (1 << bitNo)) {
					fnutil.warning(this.location(), `extra bit(s) starting with x=${x}`);
					break;
				}
				bitNo--;
			}
		}
	}

	checkLine(line) {
		if (line.match(/[^\t\f\v\u0020-\u00ff]/)) {
			throw new Error('control character(s)');
		}
		if (this.parsed.asciiChars && line.match(/[\u007f-\u00ff]/)) {
			fnutil.warning(this.location(), 'non-ascii character(s)');
		}

		switch (this.mode) {
		case MODE.META:
			if (!this.xlfdName && line.startsWith('FONT')) {
				line = this.checkFont(line);
				this.xlfdName = true;
			} else {
				this.HANDLERS.findIndex(function(handler) {
					if (line.startsWith(handler[0])) {
						handler[1](line.substring(handler[0].length).trimLeft());
						return true;
					}
					return false;
				});
			}
			break;
		case MODE.PROPS:
			if (line.startsWith('ENDPROPERTIES')) {
				this.mode = MODE.META;
			} else {
				line = this.checkProp(line);
			}
			break;
		default: // MODE.BITMAP
			if (line.startsWith('ENDCHAR')) {
				this.mode = MODE.META;
			} else {
				this.checkBitmap(line);
			}
		}
		return line;
	}

	readCheck(line, callback) {
		const match = line.match(/^COMMENT\s*bdfcheck\s+(-.*)$/);

		if (match && !this.options.parse(match[1], true, this.parsed)) {
			throw new Error('invalid bdfcheck directive');
		}

		line = callback(line);
		return line != null ? this.checkLine(line) : null;
	}

	readLines(callback) {
		return super.readLines(line => this.readCheck(line, callback));
	}
}

// -- Main --
function mainProgram(nonopt, parsed) {
	(nonopt.length >= 1 ? nonopt : [null]).forEach(input => {
		let ifs = new InputFileStream(input, parsed);

		try {
			bdf.Font.read(ifs);
			ifs.close();
		} catch (e) {
			e.message = ifs.location() + e.message;
			throw e;
		}
		ifs.proplocs.check();
		ifs.namelocs.check();
		ifs.codelocs.check();
	});
}

if (require.main === module) {
	fncli.start('bdfcheck.js', new Options(), new Params(), mainProgram);
}

File: bin/bdfcheck.py
#
# Copyright (C) 2017-2020 Dimitar Toshkov Zhekov <dimitar.zhekov@gmail.com>
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the Free
# Software Foundation; either version 2 of the License, or (at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#

import re
import codecs
from collections import OrderedDict
from enum import IntEnum, unique

import fnutil
import fncli
import fnio
import bdf

# -- Params --
class Params(fncli.Params):  # pylint: disable=too-many-instance-attributes
	def __init__(self):
		fncli.Params.__init__(self)
		self.ascii_chars = True
		self.bbx_exceeds = True
		self.dupl_codes = -1
		self.extra_bits = True
		self.attributes = True
		self.dupl_names = -1
		self.dupl_props = True
		self.common_slant = True
		self.common_weight = True
		self.xlfd_fontnm = True
		self.ywidth_zero = True


# -- Options --
HELP = ('' +
	'usage: bdfcheck [options] [INPUT...]\n' +
	'Check BDF font(s) for various problems\n' +
	'\n' +
	'  -A          disable non-ascii characters check\n' +
	'  -B          disable BBX exceeding FONTBOUNDINGBOX checks\n' +
	'  -c/-C       enable/disable duplicate character codes check\n' +
	'              (default = enabled for registry ISO10646)\n' +
	'  -E          disable extra bits check\n' +
	'  -I          disable ATTRIBUTES check\n' +
	'  -n/-N       enable duplicate character names check\n' +
	'              (default = enabled for registry ISO10646)\n' +
	'  -P          disable duplicate properties check\n' +
	'  -S          disable common slant check\n' +
	'  -W          disable common weight check\n' +
	'  -X          disable XLFD font name check\n' +
	'  -Y          disable zero WIDTH Y check\n' +
	'  --help      display this help and exit\n' +
	'  --version   display the program version and license, and exit\n' +
	'  --excstk    display the exception stack on error\n' +
	'\n' +
	'File directives: COMMENT bdfcheck --enable|disable-<check-name>\n' +
	'  (also available as long command line options)\n' +
	'\n' +
	'Check names: ascii-chars, bbx-exceeds, duplicate-codes, extra-bits,\n' +
	'  attributes, duplicate-names, duplicate-properties, common-slant,\n' +
	'  common-weight, xlfd-font, ywidth-zero\n' +
	'\n' +
	'The input BDF(s) must be v2.1 with unicode encoding.\n')

VERSION = 'bdfcheck 1.62, Copyright (C) 2017-2020 Dimitar Toshkov Zhekov\n\n' + fnutil.GPL2PLUS_LICENSE

class Options(fncli.Options):
	def __init__(self):
		fncli.Options.__init__(self, [], HELP, VERSION)


	def parse(self, name, directive, params):
		value = name.startswith('--enable') or name[1].islower()

		if name in ['-A', '--enable-ascii-chars', '--disable-ascii-chars']:
			params.ascii_chars = value
		elif name in ['-B', '--enable-bbx-exceeds', '--disable-bbx-exceeds']:
			params.bbx_exceeds = value
		elif name in ['-c', '-C', '--enable-duplicate-codes', '--disable-duplicate-codes']:
			params.dupl_codes = value
		elif name in ['-E', '--enable-extra-bits', '--disable-extra-bits']:
			params.extra_bits = value
		elif name in ['-I', '--enable-attributes', '--disable-attributes']:
			params.attributes = value
		elif name in ['-n', '-N', '--enable-duplicate-names', '--disable-duplicate-names']:
			params.dupl_names = value
		elif name in ['-P', '--enable-duplicate-properties', '--disable-duplicate-properties']:
			params.dupl_props = value
		elif name in ['-S', '--enable-common-slant', '--disable-common-slant']:
			params.common_slant = value
		elif name in ['-W', '--enable-common-weight', '--disable-common-weight']:
			params.common_weight = value
		elif name in ['-X', '--enable-xlfd-font', '--disable-xlfd-font']:
			params.xlfd_fontnm = value
		elif name in ['-Y', '--enable-ywidth-zero', '--disable-ywidth-zero']:
			params.ywidth_zero = value
		else:
			return directive is not True and self.fallback(name, params)

		return directive is not True or name.startswith('--')


# -- DupMap --
class DupMap(OrderedDict):
	def __init__(self, prefix, severity, descript, quote):
		OrderedDict.__init__(self)
		self.prefix = prefix
		self.descript = descript
		self.severity = severity
		self.quote = quote


	def check(self):
		for value, lines in self.items():
			if len(lines) > 1:
				text = 'duplicate %s %s at lines' % (self.descript, str(value))

				for index, line in enumerate(lines):
					text += ' ' if index == 0 else ' and ' if index == len(lines) - 1 else ', '
					text += str(line)

				fnutil.message(self.prefix, self.severity, text)


	def push(self, value, line_no):
		try:
			self[value].append(line_no)
		except KeyError:
			self[value] = [line_no]


# -- InputFileStream --
@unique
class MODE(IntEnum):
	META = 0
	PROPS = 1
	BITMAP = 2

class InputFileStream(fnio.InputFileStream):
	def __init__(self, file_name, parsed):
		fnio.InputFileStream.__init__(self, file_name)
		self.parsed = parsed
		self.mode = MODE.META
		self.proplocs = DupMap(self.location(), 'error', 'property', '')
		self.namelocs = DupMap(self.location(), 'warning', 'character name', '"')
		self.codelocs = DupMap(self.location(), 'warning', 'encoding', '')
		self.handlers = [
			(b'STARTCHAR', lambda value: self.append_name(value)),
			(b'ENCODING', lambda value: self.append_code(value)),
			(b'SWIDTH', lambda value: self.check_width('SWIDTH', value, bdf.Width.parse_s)),
			(b'DWIDTH', lambda value: self.check_width('DWIDTH', value, bdf.Width.parse_d)),
			(b'BBX', lambda value: self.set_last_box(value)),
			(b'BITMAP', lambda _: self.set_mode(MODE.BITMAP)),
			(b'SIZE', InputFileStream.check_size),
			(b'ATTRIBUTES', lambda value: self.check_attr(value)),
			(b'STARTPROPERTIES', lambda _: self.set_mode(MODE.PROPS)),
			(b'FONTBOUNDINGBOX', lambda value: self.set_font_box(value)),
		]
		self.xlfd_name = False
		self.last_box = None
		self.font_box = None
		self.options = Options()


	def append(self, option, valocs, value):
		if option:
			valocs.push(str(value, 'ascii'), self.line_no)


	def append_code(self, value):
		fnutil.parse_dec('encoding', value)
		self.append(self.parsed.dupl_codes, self.codelocs, value)


	def append_name(self, value):
		self.append(self.parsed.dupl_names, self.namelocs, b'"%s"' % value)


	def check_width(self, name, value, parse):
		if self.parsed.ywidth_zero and parse(name, value).y != 0:
			fnutil.warning(self.location(), 'non-zero %s Y' % name)


	def set_font_box(self, value):
		self.font_box = bdf.BBX.parse('FONTBOUNDINGBOX', value)


	def set_last_box(self, value):
		bbx = bdf.BBX.parse('BBX', value)

		if self.parsed.bbx_exceeds:
			exceeds = []

			if bbx.xoff < self.font_box.xoff:
				exceeds.append('xoff < FONTBOUNDINGBOX xoff')

			if bbx.yoff < self.font_box.yoff:
				exceeds.append('yoff < FONTBOUNDINGBOX yoff')

			if bbx.width > self.font_box.width:
				exceeds.append('width > FONTBOUNDINGBOX width')

			if bbx.height > self.font_box.height:
				exceeds.append('height > FONTBOUNDINGBOX height')

			for exceed in exceeds:
				fnutil.message(self.location(), '', exceed)

		self.last_box = bbx


	def set_mode(self, new_mode):
		self.mode = new_mode


	def check(self):
		self.process(bdf.Font.read)
		self.proplocs.check()
		self.namelocs.check()
		self.codelocs.check()


	@staticmethod
	def check_size(value):
		words = fnutil.split_words('SIZE', value, 3)
		fnutil.parse_dec('point size', words[0], 1, None)
		fnutil.parse_dec('x resolution', words[1], 1, None)
		fnutil.parse_dec('y resolution', words[2], 1, None)


	def check_attr(self, value):
		if not re.fullmatch(br'[\dA-Fa-f]{4}', value):
			raise Exception('ATTRIBUTES must be 4 hex-encoded characters')

		if self.parsed.attributes:
			fnutil.warning(self.location(), 'ATTRIBUTES may cause problems with freetype')


	def check_font(self, value):
		xlfd = value[4:].lstrip().split(b'-', 15)

		if len(xlfd) == 15 and xlfd[0] == b'':
			unicode = (xlfd[bdf.XLFD.CHARSET_REGISTRY].upper() == b'ISO10646')

			if self.parsed.dupl_codes == -1:
				self.parsed.dupl_codes = unicode

			if self.parsed.dupl_names == -1:
				self.parsed.dupl_names = unicode

			if self.parsed.common_weight:
				weight = str(xlfd[bdf.XLFD.WEIGHT_NAME], 'ascii')
				compare = weight.lower()
				consider = 'Bold' if 'bold' in compare else 'Normal'

				if compare in ['medium', 'regular']:
					compare = 'normal'

				if compare != consider.lower():
					fnutil.warning(self.location(), 'weight "%s" may be considered %s' % (weight, consider))

			if self.parsed.common_slant:
				slant = str(xlfd[bdf.XLFD.SLANT], 'ascii')
				consider = 'Italic' if re.search('^[IO]', slant) else 'Regular'

				if not re.fullmatch('[IOR]', slant):
					fnutil.warning(self.location(), 'slant "%s" may be considered %s' % (slant, consider))

		else:
			if self.parsed.xlfd_fontnm:
				fnutil.warning(self.location(), 'non-XLFD font name')

			value = b'FONT --------------'

		return value


	def check_prop(self, line):
		match = re.fullmatch(br'(\w+)\s+([-\d"].*)', line)

		if not match:
			raise Exception('invalid property format')

		name = match.group(1)
		value = match.group(2)

		if value.startswith(b'"'):
			if len(value) < 2 or not value.endswith(b'"'):
				raise Exception('no closing double quote')
			if re.search(b'[^"]"[^"]', value[1 : len(value) - 1]):
				raise Exception('unescaped double quote')
		else:
			fnutil.parse_dec('value', value, None, None)

		self.append(self.parsed.dupl_props, self.proplocs, name)
		return b'P%d 1' % self.line_no


	def check_bitmap(self, line):
		if len(line) != self.last_box.row_size() * 2:
			raise Exception('invalid bitmap length')

		data = codecs.decode(line, 'hex')

		if self.parsed.extra_bits:
			check_x = (self.last_box.width - 1) | 7
			last_byte = data[len(data) - 1]
			bit_no = 7 - (self.last_box.width & 7)

			for x in range(self.last_box.width, check_x + 1):
				if last_byte & (1 << bit_no):
					fnutil.warning(self.location(), 'extra bit(s) starting with x=%d' % x)
					break
				bit_no -= 1


	def check_line(self, line):
		if re.search(b'[^\t\f\v\x20-\xff]', line):
			raise Exception('control character(s)')

		if self.parsed.ascii_chars and re.search(b'[\x7f-\xff]', line):
			fnutil.warning(self.location(), 'non-ascii character(s)')

		if self.mode == MODE.META:
			if not self.xlfd_name and line.startswith(b'FONT'):
				line = self.check_font(line)
				self.xlfd_name = True
			else:
				for handler in self.handlers:
					if line.startswith(handler[0]):
						handler[1](line[len(handler[0]):].lstrip())
						break
		elif self.mode == MODE.PROPS:
			if line.startswith(b'ENDPROPERTIES'):
				self.mode = MODE.META
			else:
				line = self.check_prop(line)
		else:  # MODE.BITMAP
			if line.startswith(b'ENDCHAR'):
				self.mode = MODE.META
			else:
				self.check_bitmap(line)

		return line


	def read_check(self, line, callback):
		match = re.search(br'^COMMENT\s*bdfcheck\s+(-.*)$', line)

		if match and not self.options.parse(str(match[1], 'ascii'), True, self.parsed):
			raise Exception('invalid bdfcheck directive')

		line = callback(line)
		return self.check_line(line) if line is not None else None


	def read_lines(self, callback):
		return fnio.InputFileStream.read_lines(self, lambda line: self.read_check(line, callback))


# -- Main --
def main_program(nonopt, parsed):
	for input_name in nonopt or [None]:
		InputFileStream(input_name, parsed).check()


if __name__ == '__main__':
	fncli.start('bdfcheck.py', Options(), Params(), main_program)
